### PR TITLE
[MRG]: Add cell parameters in GUI

### PR DESCRIPTION
**Pull request by kmilo9999**
This is the work for the cell params feature for L2 and L5 pyramidal cell types
cell_params

### CODE REVIEW

File: hnn_core/gui/gui.py [hunk 1]
```diff
@@ -32,6 +32,81 @@
 import numpy as np
 from copy import deepcopy
 
+cell_parameters_dic = {
+
+    "Geometry":
+    [
+        ('Soma lenght', 'micron'),
```
Comment (ntolley) [inline hnn_core/gui/gui.py:39]:
```suggestion
        ('Soma length', 'micron'),
```

Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Done

File: hnn_core/gui/gui.py [hunk 2]
```diff
@@ -491,19 +610,32 @@ def compose(self, return_layout=True):
...
+
         # accordions to group local-connectivity by cell type
-        connectivity_boxes = [
-            VBox(slider) for slider in self.connectivity_widgets]
-        connectivity_names = (
-            'Layer 2/3 Pyramidal', 'Layer 5 Pyramidal', 'Layer 2 Basket',
-            'Layer 5 Basket')
-        cell_connectivity = Accordion(children=connectivity_boxes)
-        cell_connectivity.titles = [s for s in connectivity_names]
+        # connectivity_boxes = [
```
Comment (ntolley) [inline hnn_core/gui/gui.py:632]: delete blocks commented out?

File: hnn_core/gui/gui.py [hunk 3]
```diff
@@ -1223,7 +1375,68 @@ def add_connectivity_tab(params, connectivity_out,
...
+                                     cell_pameters_vboxes,
+                                     cell_type_radio_button.value,
+                                     cell_layer_radio_button.value))
+
+
+def get_cell_param_default_value(cell_type_key, param_dict):
+    return param_dict[cell_type_key]
+
+
+def get_cell_params_dic_values(network_cell_types):
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1431]: seems like this function isn't being used anywhere

File: hnn_core/gui/gui.py [hunk 4]
```diff
@@ -1223,7 +1375,68 @@ def add_connectivity_tab(params, connectivity_out,
...
+def add_cell_parameters_tab(network, cell_params_out, cell_pameters_vboxes,
+                            cell_layer_radio_button, cell_type_radio_button,
+                            layout):
+    L2_defautl_values = get_L2Pyr_params_default()
+    L5_default_values = get_L5Pyr_params_default()
+    cell_types = [("L2", L2_defautl_values), ("L5", L5_default_values)]
+    style = {'description_width': '235px'}
+    kwargs = dict(layout=layout, style=style)
+
+    # get_cell_params_dic_values(network.cell_types)
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1390]:
```suggestion
```

File: hnn_core/gui/gui.py [hunk 5]
```diff
@@ -1594,5 +1991,9 @@ def launch():
     You can pass voila commandline parameters as usual.
     """
     from voila.app import main
+    import debugpy
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1994]: Is this specific for your setup to have the VSCode debugger activate inside the GUI?

Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Yeah, I forgot to remove it

File: hnn_core/gui/gui.py (hunk 2 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Done

File: hnn_core/gui/gui.py (hunk 3 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: I removed it

File: hnn_core/gui/gui.py (hunk 4 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: I Removed it

File: hnn_core/gui/gui.py [hunk 6]
```diff
@@ -1223,7 +1366,57 @@ def add_connectivity_tab(params, connectivity_out,
...
         display(cell_connectivity)
 
-    return net
+    return network
+
+
+def add_cell_parameters_tab(cell_params_out, cell_pameters_vboxes,
+                            cell_layer_radio_button, cell_type_radio_button,
+                            layout):
+    L2_defautl_values = get_L2Pyr_params_default()
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:1375]:
```suggestion
    L2_default_values = get_L2Pyr_params_default()
```

File: hnn_core/gui/gui.py [hunk 7]
```diff
@@ -1516,6 +1742,160 @@ def run_button_clicked(widget_simulation_name, log_out, drive_widgets,
...
+
+    # Dentrite
+    for section, indices in zip(dendrite_sections, param_indices):
+        sections[section]._L = cell_params[indices[0]].value
+        sections[section]._diam = cell_params[indices[1]].value
+        sections[section]._cm = dendrite_cm
+        sections[section]._Ra = dendrite_Ra
+
+
+def update_synapse_cell_params(net, cell_param_key, param_list):
```
Comment (jasmainak) [inline hnn_core/gui/gui.py:1788]: private function since it's not documented and may change in the future ...

File: hnn_core/gui/gui.py (hunk 6 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Done

File: hnn_core/gui/gui.py [hunk 8]
```diff
@@ -284,6 +364,14 @@ def __init__(self, theme_color="#8A2BE2",
             'Delete drives', 'success', layout=self.layout['btn'],
             button_color=self.layout['theme_color'])
 
+        self.cell_type_radio_buttons = RadioButtons(
+            options=['L2 Pyramidal', 'L5 Pyramidal'],
+            description='Cell type:')
+
+        self.cell_layer_radio_buttons = RadioButtons(
+            options=['Geometry', 'Synapses', 'Biophysics'],
+            description='Layer:')
```
Comment (ntolley) [inline hnn_core/gui/gui.py:373]:
```suggestion
            description='Cell properties:')
```
?

File: hnn_core/gui/gui.py [hunk 9]
```diff
@@ -1187,17 +1332,17 @@ def add_connectivity_tab(params, connectivity_out,
                 # the connectivity list should be built on this level
                 receptor_related_conn = {}
                 for receptor in receptors:
-                    conn_indices = pick_connection(net=net,
+                    conn_indices = pick_connection(net=network,
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1335]: why rename to `network`? nitpick but by convention we use `net` throughout the code base

File: hnn_core/gui/gui.py (hunk 7 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py:1788]: renamed as a private function

File: hnn_core/gui/gui.py (hunk 8 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Changed the description as indicated

File: hnn_core/gui/gui.py (hunk 9 above)
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: renamed back to `net`

File: doc/gui/tutorial_erp.ipynb [hunk 10]
```diff
@@ -220,7 +234,7 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "gui._simulate_left_tab_click(\"Network connectivity\")"
+    "gui._simulate_left_tab_click(\"Network\")"
```
Comment (ntolley) [inline doc/gui/tutorial_erp.ipynb:237]: I think we're going to move away from the automatically generated GUI documentation so not necessary to update this file (no need to remove this change of course)

**Comment by ntolley** [comment]
Looks like the test is failing here:
image

File: hnn_core/gui/gui.py [hunk 11]
```diff
@@ -1516,6 +1741,160 @@ def run_button_clicked(widget_simulation_name, log_out, drive_widgets,
...
+    sections['soma']._L = cell_params[0].value
+    sections['soma']._diam = cell_params[1].value
+    sections['soma']._cm = cell_params[2].value
+    sections['soma']._Ra = cell_params[3].value
+
+    # Dendrite common parameters
+    dendrite_cm = cell_params[4].value
+    dendrite_Ra = cell_params[5].value
+
+    dendrite_sections = [
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1771]:
```suggestion
    dendrite_sections = [name for name in sections.keys() if name != 'soma']
```
This list is actually missing `apical_2` so perhaps it's better to just loop over the section names

File: hnn_core/gui/gui.py [hunk 12]
```diff
@@ -1516,6 +1741,160 @@ def run_button_clicked(widget_simulation_name, log_out, drive_widgets,
...
+        'apical_trunk', 'apical_1', 'apical_tuft', 'apical_oblique',
+        'basal_1', 'basal_2', 'basal_3'
+    ]
```
Comment (ntolley) [inline hnn_core/gui/gui.py:1774]:
```suggestion
    dendrite_sections = [name for name in sections.keys() if name != 'soma']
```
This list is actually missing `apical_2` so perhaps it's better to just loop over the section names

Comment (kmilo9999) [reply hnn_core/gui/gui.py]: Done

File: doc/gui/tutorial_erp.ipynb (hunk 10 above)
Comment (kmilo9999) [reply doc/gui/tutorial_erp.ipynb:237]: I think one of the tests fails if the 'Network Connectivity' tab title changes but it's not updated in the notebook.

File: hnn_core/gui/gui.py [hunk 13]
```diff
@@ -491,19 +610,23 @@ def compose(self, return_layout=True):
...
             self._connectivity_out,
         ])
 
-        # accordions to group local-connectivity by cell type
-        connectivity_boxes = [
-            VBox(slider) for slider in self.connectivity_widgets]
-        connectivity_names = (
-            'Layer 2/3 Pyramidal', 'Layer 5 Pyramidal', 'Layer 2 Basket',
-            'Layer 5 Basket')
-        cell_connectivity = Accordion(children=connectivity_boxes)
```
Comment (ntolley) [inline hnn_core/gui/gui.py:505]: Was this somehow redundant or being unused? I see that you're using `add_connectivity_tab` to setup the connectivity, but that function already existed before this PR

Comment (kmilo9999) [reply hnn_core/gui/gui.py:505]: I dont think it was being used at all. That variable only exists during the scope of that function.

**Review by ntolley — APPROVED** [review]
This is all good on my end @kmilo9999, really amazing work here!!

@gtdang feel free to merge if everything looks good to you

File: hnn_core/gui/gui.py [hunk 14]
```diff
@@ -26,12 +26,92 @@
...
 from hnn_core.dipole import _read_dipole_txt
+from hnn_core.params_default import (get_L2Pyr_params_default,
+                                     get_L5Pyr_params_default)
 
 import base64
 import zipfile
 import numpy as np
 from copy import deepcopy
 
+cell_parameters_dict = {
```
Comment (gtdang) [inline hnn_core/gui/gui.py:37]: A suggestion for a future PR... Changing these to dictionaries with index at the keys and the tuples as the values would help with configuring and debugging the parameter mappings in the `update_*_cell_params` functions assuming there is a 1:1 order mapping of widgets to these lists.

**Review by gtdang — APPROVED** [review]
Have some minor comments but otherwise approve.

File: hnn_core/gui/gui.py [hunk 15]
```diff
@@ -1360,21 +1564,40 @@ def _init_network_from_widgets(params, dt, tstop, single_simulation_data,
...
+        'Synapses': _update_synapse_cell_params,
+        'L2 Pyramidal_Biophysics': _update_L2_biophysics_cell_params,
+        'L5 Pyramidal_Biophysics': _update_L5_biophysics_cell_params
+    }
+
+    # Update cell params
+    for vbox_key, cell_param_list in cell_params_vboxes.items():
+        for key, update_function in update_functions.items():
+            if key in vbox_key:
+                # Remove 'Pyramidal' keyword
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply hnn_core/gui/gui.py]: I removed it.

File: hnn_core/tests/test_gui.py [hunk 16]
```diff
@@ -657,6 +659,41 @@ def test_gui_add_tonic_input(setup_gui):
...
+    for pyramid_cell_type in pyramid_cell_types:
+        cell_type = pyramid_cell_type.split('_')[0]
+        for cell_layer in layers:
+            key = f'{cell_type} Pyramidal_{cell_layer}'
+            assert (any(key in k for k in keys))
+            num_cell_params = num_cell_params + 1
+
+    assert (len(keys) == num_cell_params)
+
+    # Check the if the cell params dictionary has been updated
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply hnn_core/tests/test_gui.py:698]: Done

File: hnn_core/tests/test_gui.py [hunk 17]
```diff
@@ -657,6 +659,41 @@ def test_gui_add_tonic_input(setup_gui):
...
+    assert (len(layers) == 3)
+
+    keys = gui.cell_pameters_widgets.keys()
+    num_cell_params = 0
+    for pyramid_cell_type in pyramid_cell_types:
+        cell_type = pyramid_cell_type.split('_')[0]
+        for cell_layer in layers:
+            key = f'{cell_type} Pyramidal_{cell_layer}'
+            assert (any(key in k for k in keys))
+            num_cell_params = num_cell_params + 1
```
[the remark this answers is not in the document]
Comment (kmilo9999) [reply hnn_core/tests/test_gui.py]: Done

**Comment by gtdang** [comment]
Amazing job @kmilo9999!

**Pull request merged**